### PR TITLE
[EC-261] SCIM

### PR DESCRIPTION
**Pull request by cscharf**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Adds a new SCIM service which accepts callbacks from IdPs as a SCIM provider for Bitwarden. Also includes hooks for a new type of API Key and Organization Connection around SCIM specifically with slightly different requirements than others.

## Code changes
Client PR: https://github.com/bitwarden/clients/pull/3065

A note on the EF migrations for pgsql and mysql - with the dotnet version update, EF decided to take on several different changes automatically on how it applied `DateTime` types within the pgsql 6 drivers. Please see [this article](https://www.roji.org/postgresql-dotnet-timestamp-mapping) on the context of adding the switch, `"Npgsql.EnableLegacyTimestampBehavior", true`

## Before you submit

```
- [X] I have checked for formatting errors (`dotnet format --verify-no-changes`) (required)
- [X] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [X] This change requires a **documentation update** (notify the documentation team)
- [X] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

File: src/Core/Models/OrganizationConnectionConfigs/ScimConfig.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Text.Json.Serialization;
+using Bit.Core.Enums;
+
+namespace Bit.Core.Models.OrganizationConnectionConfigs
+{
+    public class ScimConfig
+    {
+        public bool Enabled { get; set; }
```
Comment (justindbaur) [inline src/Core/Models/OrganizationConnectionConfigs/ScimConfig.cs:8]: There is a `Enabled` column on the table level for this. Does there also need to be one in the config?

File: util/PostgresMigrations/Scripts/2022-05-24_00_DeviceUnknownVerification.psql [hunk 2]
```diff
@@ -1,6 +1,6 @@
 START TRANSACTION;
 
-ALTER TABLE "User" ADD "UnknownDeviceVerificationEnabled" boolean NOT NULL DEFAULT 1;
+ALTER TABLE "User" ADD "UnknownDeviceVerificationEnabled" boolean NOT NULL DEFAULT true;
```
Comment (justindbaur) [inline util/PostgresMigrations/Scripts/2022-05-24_00_DeviceUnknownVerification.psql:3]: Nice fix.

**Review by justindbaur — CHANGES_REQUESTED** [review]

File: bitwarden_license/src/Scim/Controllers/v2/GroupsController.cs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+using Bit.Scim.Context;
+using Bit.Scim.Models;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Options;
+
+namespace Bit.Scim.Controllers.v2
+{
+    [Authorize("Scim")]
+    [Route("v2/{organizationId}/groups")]
```
[the remark this answers is not in the document]
Comment (cscharf) [reply bitwarden_license/src/Scim/Controllers/v2/GroupsController.cs:13]: That's the protocol version of the SCIM standard

File: src/Core/Models/OrganizationConnectionConfigs/ScimConfig.cs (hunk 1 above)
Comment (cscharf) [reply src/Core/Models/OrganizationConnectionConfigs/ScimConfig.cs:8]: According to @MGibson1 , the `Enabled` flag was for a different purpose so I created a lower-level flag for the purpose of SCIM itself, if we need to flatten that back out and push the connection up further we can do that and maybe simplify things a bit.

File: bitwarden_license/src/Scim/Controllers/v2/UsersController.cs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            if (orgUserByExternalId != null)
+            {
+                return new ConflictResult();
+            }
+
+            var invitedOrgUser = await _organizationService.InviteUserAsync(organizationId, null, email,
+                OrganizationUserType.User, false, externalId, new List<SelectionReadOnly>());
+            var orgUser = await _organizationUserRepository.GetDetailsByIdAsync(invitedOrgUser.Id);
+            var response = new ScimUserResponseModel(orgUser);
+            return new CreatedResult(Url.Action(nameof(Get), new { orgUser.OrganizationId, orgUser.Id }), response);
```
[the remark this answers is not in the document]
Comment (cscharf) [reply bitwarden_license/src/Scim/Controllers/v2/UsersController.cs:178]: That's all Kyle 😁

File: bitwarden_license/src/Scim/Controllers/v2/UsersController.cs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                    {
+                        await _organizationService.DeactivateUserAsync(orgUser, null);
+                        operationHandled = true;
+                    }
+                }
+            }
+
+            if (!operationHandled)
+            {
+                _logger.LogWarning("User patch operation not handled: {0} : ",
```
[the remark this answers is not in the document]
Comment (cscharf) [reply bitwarden_license/src/Scim/Controllers/v2/UsersController.cs:244]: Done, thanks!

File: bitwarden_license/src/Scim/Scim.csproj [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <UserSecretsId>bitwarden-Scim</UserSecretsId>
```
[the remark this answers is not in the document]
Comment (cscharf) [reply bitwarden_license/src/Scim/Scim.csproj:4]: Good call, thanks!

File: bitwarden_license/src/Scim/Utilities/Constants.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Bit.Scim.Utilities
+{
+    public static class Constants
```
[the remark this answers is not in the document]
Comment (cscharf) [reply bitwarden_license/src/Scim/Utilities/Constants.cs:3]: I agree it _seems_ redundant, however it is namespaced and it's very common practice to have consistent naming across namespaces. I'll leave this one to @kspearrin however.

[the remark this answers is not in the document]
Comment (kspearrin) [reply bitwarden_license/src/Scim/Utilities/Constants.cs:3]: Done

**Review by kspearrin — APPROVED** [review]

File: bitwarden_license/src/Scim/Utilities/Constants.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (justindbaur) [reply bitwarden_license/src/Scim/Utilities/Constants.cs]: I always appreciate their being one file of a name so I can more blindly navigate around in VSCode.

File: src/Core/Models/OrganizationConnectionConfigs/ScimConfig.cs (hunk 1 above)
Comment (justindbaur) [reply src/Core/Models/OrganizationConnectionConfigs/ScimConfig.cs:8]: That's fine, I will defer to him.

**Review by justindbaur — APPROVED** [review]
Looks good!

**Pull request merged**